qiskit_experiments/curve_analysis/scatter_table.py

```diff
@@ -9,185 +9,498 @@
 # Any modifications or derivative works of this code must retain this
 # copyright notice, and modified files need to carry a notice indicating
 # that they have been altered from the originals.
+
 """Table representation of the x, y data for curve fitting."""
+from __future__ import annotations
 
 import logging
-from typing import List, Sequence, Dict, Any, Union
+import warnings
+from collections.abc import Iterator
+from typing import Any
+from itertools import groupby
+from operator import itemgetter
 
 import numpy as np
 import pandas as pd
 
 from qiskit.utils import deprecate_func
 
-from qiskit_experiments.framework.table_mixin import DefaultColumnsMixIn
-
 
 LOG = logging.getLogger(__name__)
 
 
-class ScatterTable(pd.DataFrame, DefaultColumnsMixIn):
-    """A table to store x and y data with metadata associated with the data point.
+class ScatterTable:
+    """A table-like dataset for curve fitting intermediate data.
+
+    Default table columns are defined in the class attribute :attr:`.DEFAULT_COLUMNS`.
+    This table cannot be expanded with user-provided column names.
+
+    This dataset is not thread safe. Do not use the same instance in multiple threads.
+
+    .. _filter_scatter_table:
+
+    Filtering ScatterTable
+    ----------------------
+
+    ScatterTable is a single source of the truth as the data used in the curve fit analysis.
+    Each data point in a 1-D curve fit may consist of the x value, y value, and
+    standard error of the y value.
+    In addition, such analysis may internally create several data subsets,
+    and data points can also take metadata triplet (`data_uid`, `category`, `analysis`)
+    to distinguish the subset.
+
+    * The `data_uid` is an integer key representing the class of the data.
```

Comment:
Thank you for adding this info in the doc, @nkanazawa1989 . It helps me a lot to understand the concepts to know for users of ScatterTable class. I've now understood `data_uid` identifies a "series" of data to fit a model. And `name` is another human-friendly identifier that indicates a "series" of data to fit a model. Given that, `series_id` and `series_name` sound better to me. What do you think? @nkanazawa1989 @wshanks

Reply:
`series_index` instead of `series_id`? Looking at the example in the tutorial, `data_uid` can take a common value for different category. In that sense, it's not "ID" (to me, ID sounds like a solo unique identifier), it's more like "index" defined for each (`category`, `analysis`) data.

Reply:
I like this suggestion. Perhaps `series_idx` to be shorter?

Reply:
One case to consider regarding the naming is `StarkRamseyXYAmpScanAnalysis`. For that experiment, X and Y vs amp data are recorded. Then the X and Y data are converted into phase vs amp data which is fit. So there are two raw series that become the formatted series that is fit. Maybe `series_` still makes sense since one series is passed to the model. It is the fit models though that have a one to one mapping the `data_uid` values.

Reply:
Umm, I'm converging to Will's comment:
> One concern I have to make the roles of name, data_uid, category, and model (not a column but a term used in the same context) clear.

Maybe, all we need is `model_` (we can distinguish "ramsey_xy" and "freq" using `model_name` instead of `category`).

Reply:
We cannot define fit model for ramsey_xy because its P1 envelope is unpredictable. In this sense, "ramsey_xy" is another the intermediate data to produce the frequency curves which have fit models (Ramsey data is used for visualization and we cannot discard -- when fit doesn't go well experimentalists can get some insights from the raw Ramsey signal). This is why I cannot introduce fixed categories -- this requires the Ramsey analysis to define ScatterTable subclass with its own categories.

Reply:
Anyways I also like series. This was originally used in the curve analysis to indicate the data belongs to a particular fit model.

Reply:
Done in [f9d13bd](https://github.com/Qiskit-Extensions/qiskit-experiments/pull/1360/commits/f9d13bd59b27deab0e49a0f535dde3375bb051b0). Thanks for the suggestion.

Reply:
So I think the broadest way to think about the columns are that `series` labels a single curve of some kind (`ScatterTable` is under `qiskit_experiments.curve_analysis` so it is safe to assume all the data belong to curves). `category` is a label for a group of curves/series that all belong to the same stage of processing (raw/formatted/fitted). `analysis` is a namespace for separating these groups of curves from different subexperiments.

@coruscating Is it important to get the release out today? I think I would like to clarify these terms a bit in the docs, but that could be done in a follow up if it is important to release today. I don't have any objections to the code (as long as further discussion doesn't make us want to change any names again).